media/Jsontracer/ruler.js

```diff
@@ -0,0 +1,103 @@
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//    * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//    * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//    * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// This file referenced the result of
+// https://github.com/catapult-project/catapult/tree/444aba89e1c30edf348c611a9df79e2376178ba8/tracing
+
+export default function renderRuler(timeLimit, digit) {
+  const graph = document.querySelector('.graph');
+
+  const rulerContainer = document.createElement('div');
+  rulerContainer.className = 'ruler-container';
+
+  const rulerBlank = document.createElement('div');
+  rulerBlank.className = 'ruler-blank';
+
+  const ruler = document.createElement('div');
+  ruler.className = 'ruler';
+
+  rulerContainer.append(rulerBlank, ruler);
+  graph.append(rulerContainer);
+
+  mapToRulergraduation(timeLimit, digit);
+}
+
+function mapToRulergraduation(timeLimit, digit) {
+  const ruler = document.querySelector('.ruler');
+
+  for (let i = 0; i < parseInt(timeLimit / 10 ** (digit - 1)); i++) {
+    const graduation = document.createElement('div');
+    graduation.className = 'graduation';
+
+    for (let j = 0; j < 5; j++) {
+      const smallGraduation = document.createElement('div');
+      smallGraduation.className = 'small-graduation';
+
+      if (j === 0) {
+        const index = document.createElement('div');
+        index.className = 'index';
+        index.innerText = calculateGraduation(i * 10 ** (digit - 1));
+        smallGraduation.append(index);
+      }
+
+      graduation.append(smallGraduation);
+    }
+
+    ruler.append(graduation);
+  }
+}
+
+function calculateGraduation(graduation) {
```

Comment:
I saw exactly same code in `media/Jsontracer/dynamicGraduation.js`.
Could you add `TODO Reduce duplicate function` ?

Reply:
OK, i will add `TODO` and apply that to the next PR : )

Reply:
Since this is not that urgent, you can do it later :D